presto-cache/src/main/java/com/facebook/presto/cache/LocalRangeCacheManager.java

```diff
@@ -0,0 +1,461 @@
+        this.cache = CacheBuilder.newBuilder()
+                .maximumSize(cacheConfig.getMaxCachedEntries())
+                .expireAfterAccess(cacheConfig.getCacheTtl().toMillis(), MILLISECONDS)
+                .removalListener(new CacheRemovalListener())
+                .recordStats()
+                .build();
+        this.stats = requireNonNull(stats, "stats is null");
+        this.directory = new Path(cacheConfig.getBaseDirectory());
+        checkArgument(cacheConfig.getMaxInMemoryCacheSize().toBytes() >= 0, "maxInflightBytes is negative");
+        this.maxInflightBytes = cacheConfig.getMaxInMemoryCacheSize().toBytes();
+
+        File target = new File(directory.toUri());
+        if (!target.exists()) {
+            try {
+                Files.createDirectories(target.toPath());
+            }
+            catch (IOException e) {
+                throw new PrestoException(GENERIC_INTERNAL_ERROR, "cannot create cache directory " + target, e);
+            }
+        }
+        else {
+            File[] files = target.listFiles();
+            if (files == null) {
+                return;
+            }
+
+            this.cacheRemovalExecutor.submit(() -> Arrays.stream(files).forEach(file -> {
+                try {
+                    Files.delete(file.toPath());
+                }
+                catch (IOException e) {
+                    // ignore
+                }
+            }));
+        }
+    }
+
+    @PreDestroy
+    public void destroy()
+    {
+        cacheFlushExecutor.shutdownNow();
+        cacheRemovalExecutor.shutdownNow();
+    }
+
+    @Override
+    public boolean get(FileCacheKey key, byte[] buffer, int offset)
+    {
+        boolean result = read(key, buffer, offset);
+        if (result) {
+            stats.incrementCacheHit();
+        }
+        else {
+            stats.incrementCacheMiss();
+        }
+
+        return result;
+    }
+
+    @Override
+    public void put(FileCacheKey key, Slice data)
+    {
+        if (stats.getInMemoryRetainedBytes() + data.length() >= maxInflightBytes) {
+            // cannot accept more requests
+            return;
+        }
+
+        // make a copy given the input data could be a reusable buffer
+        stats.addInMemoryRetainedBytes(data.length());
+        byte[] copy = data.getBytes();
+        cacheFlushExecutor.submit(() -> {
+            Path newPathHint = new Path(directory.toUri() + "/" + randomUUID() + EXTENSION);
+            if (!write(key, copy, newPathHint)) {
+                log.warn("%s Fail to persist cache %s with length %s ", Thread.currentThread().getName(), newPathHint, key.getLength());
+            }
+            stats.addInMemoryRetainedBytes(-copy.length);
+        });
+    }
+
+    private boolean read(FileCacheKey key, byte[] buffer, int offset)
+    {
+        if (key.getLength() <= 0) {
+            // no-op
+            return true;
+        }
+
+        try {
+            cache.get(key.getPath(), () -> true);
```

Comment:
Curious: Are we trying to refresh "last accessed" clock here?

Reply:
Yep, it is to hint the cache